File: binding/src/libquuid.cpp
#include <QByteArray>
#include <QByteArrayView>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QUuid>
#define WORKAROUND_INNER_CLASS_DEFINITION_QUuid__Id128Bytes
#include <quuid.h>
#include "libquuid.h"
#include "libquuid.hxx"

_GUID* _GUID_new() {
	 return new _GUID();
}

_GUID* _GUID_new2(const _GUID* param1) {
	 return new _GUID(*param1);
}

unsigned long _GUID_Data1(const _GUID* self) {
	return self->Data1;
}

void _GUID_SetData1(_GUID* self, unsigned long Data1) {
	self->Data1;
}

uint16_t _GUID_Data2(const _GUID* self) {
	return self->Data2;
}

void _GUID_SetData2(_GUID* self, uint16_t Data2) {
	self->Data2;
}

uint16_t _GUID_Data3(const _GUID* self) {
	return self->Data3;
}

void _GUID_SetData3(_GUID* self, uint16_t Data3) {
	self->Data3;
}

void _GUID_Delete(_GUID* self) {
    delete self;
}

QUuid* QUuid_new(const QUuid* other) {
	 return new QUuid(*other);
}

QUuid* QUuid_new2(QUuid* other) {
	 return new QUuid(*other);
}

QUuid* QUuid_new3() {
	 return new QUuid();
}

QUuid* QUuid_new4(unsigned int l, uint16_t w1, uint16_t w2, unsigned char b1, unsigned char b2, unsigned char b3, unsigned char b4, unsigned char b5, unsigned char b6, unsigned char b7, unsigned char b8) {
	 return new QUuid(l, w1, w2, b1, b2, b3, b4, b5, b6, b7, b8);
}

QUuid* QUuid_new5(QUuid__Id128Bytes* id128) {
	 return new QUuid(*id128);
}

QUuid* QUuid_new6(const _GUID* guid) {
	 return new QUuid(*guid);
}

QUuid* QUuid_new7(const QUuid* param1) {
	 return new QUuid(*param1);
}

QUuid* QUuid_new8(QUuid__Id128Bytes* id128, int order) {
	 return new QUuid(*id128, static_cast<QSysInfo::Endian>(order));
}

void QUuid_CopyAssign(QUuid* self, QUuid* other) {
    *self = *other;
}

void QUuid_MoveAssign(QUuid* self, QUuid* other) {
    *self = std::move(*other);
}

libqt_string QUuid_ToString(const QUuid* self) {
	QString _ret = self->toString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QUuid_ToByteArray(const QUuid* self) {
	QByteArray _qb = self->toByteArray();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QUuid__Id128Bytes* QUuid_ToBytes(const QUuid* self) {
	return new QUuid::Id128Bytes(self->toBytes());
}

libqt_string QUuid_ToRfc4122(const QUuid* self) {
	QByteArray _qb = self->toRfc4122();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QUuid* QUuid_FromBytes(const void* bytes) {
	return new QUuid(QUuid::fromBytes(bytes));
}

QUuid* QUuid_FromRfc4122(QByteArrayView* param1) {
	return new QUuid(QUuid::fromRfc4122(*param1));
}

bool QUuid_IsNull(const QUuid* self) {
	return self->isNull();
}

void QUuid_OperatorAssign(QUuid* self, const _GUID* guid) {
	self->operator=(*guid);
}

_GUID* QUuid_Operator_GUID(const QUuid* self) {
	return new _GUID(self->operator _GUID());
}

QUuid* QUuid_CreateUuid() {
	return new QUuid(QUuid::createUuid());
}

QUuid* QUuid_CreateUuidV5(QUuid* ns, QByteArrayView* baseData) {
	return new QUuid(QUuid::createUuidV5(*ns, *baseData));
}

QUuid* QUuid_CreateUuidV3(QUuid* ns, QByteArrayView* baseData) {
	return new QUuid(QUuid::createUuidV3(*ns, *baseData));
}

int QUuid_Variant(const QUuid* self) {
	return self->variant();
}

int QUuid_Version(const QUuid* self) {
	return self->version();
}

unsigned int QUuid_Data1(const QUuid* self) {
	return self->data1;
}

void QUuid_SetData1(QUuid* self, unsigned int data1) {
	self->data1;
}

uint16_t QUuid_Data2(const QUuid* self) {
	return self->data2;
}

void QUuid_SetData2(QUuid* self, uint16_t data2) {
	self->data2;
}

uint16_t QUuid_Data3(const QUuid* self) {
	return self->data3;
}

void QUuid_SetData3(QUuid* self, uint16_t data3) {
	self->data3;
}

void QUuid_OperatorAssign2(QUuid* self, const QUuid* param1) {
	self->operator=(*param1);
}

libqt_string QUuid_ToString1(const QUuid* self, int mode) {
	QString _ret = self->toString(static_cast<QUuid::StringFormat>(mode));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QUuid_ToByteArray1(const QUuid* self, int mode) {
	QByteArray _qb = self->toByteArray(static_cast<QUuid::StringFormat>(mode));
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QUuid__Id128Bytes* QUuid_ToBytes1(const QUuid* self, int order) {
	return new QUuid::Id128Bytes(self->toBytes(static_cast<QSysInfo::Endian>(order)));
}

QUuid* QUuid_FromBytes2(const void* bytes, int order) {
	return new QUuid(QUuid::fromBytes(bytes, static_cast<QSysInfo::Endian>(order)));
}

void QUuid_Delete(QUuid* self) {
    delete self;
}

QUuid__Id128Bytes* QUuid__Id128Bytes_new() {
	 return new QUuid::Id128Bytes();
}

QUuid__Id128Bytes* QUuid__Id128Bytes_new2(const QUuid__Id128Bytes* param1) {
	 return new QUuid::Id128Bytes(*param1);
}

QByteArrayView* QUuid__Id128Bytes_OperatorQByteArrayView(const QUuid__Id128Bytes* self) {
	return new QByteArrayView(self->operator QByteArrayView());
}

void QUuid__Id128Bytes_Delete(QUuid__Id128Bytes* self) {
    delete self;
}

File: binding/include/libquuid.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/_guid.html
pub const _guid = struct {

    /// New constructs a new _GUID object.
    pub fn New() QtC._GUID {
        return qtc._GUID_new();
    }


    /// New2 constructs a new _GUID object.
    pub fn New2(param1: ?*anyopaque) QtC._GUID {
        return qtc._GUID_new2(@ptrCast(param1));
    }


    pub fn Data1(self: ?*anyopaque, ) unsigned long {
        return qtc._GUID_Data1(@ptrCast(self));
    }

    pub fn SetData1(self: ?*anyopaque, Data1: unsigned long) void {
        qtc._GUID_SetData1(@ptrCast(self), Data1);
    }

    pub fn Data2(self: ?*anyopaque, ) u16 {
        return qtc._GUID_Data2(@ptrCast(self));
    }

    pub fn SetData2(self: ?*anyopaque, Data2: u16) void {
        qtc._GUID_SetData2(@ptrCast(self), Data2);
    }

    pub fn Data3(self: ?*anyopaque, ) u16 {
        return qtc._GUID_Data3(@ptrCast(self));
    }

    pub fn SetData3(self: ?*anyopaque, Data3: u16) void {
        qtc._GUID_SetData3(@ptrCast(self), Data3);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc._GUID_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/quuid.html
pub const quuid = struct {

    /// New constructs a new QUuid object.
    pub fn New(other: ?*anyopaque) QtC.QUuid {
        return qtc.QUuid_new(@ptrCast(other));
    }


    /// New2 constructs a new QUuid object.
    pub fn New2(other: ?*anyopaque) QtC.QUuid {
        return qtc.QUuid_new2(@ptrCast(other));
    }


    /// New3 constructs a new QUuid object.
    pub fn New3() QtC.QUuid {
        return qtc.QUuid_new3();
    }


    /// New4 constructs a new QUuid object.
    pub fn New4(l: u32, w1: u16, w2: u16, b1: u8, b2: u8, b3: u8, b4: u8, b5: u8, b6: u8, b7: u8, b8: u8) QtC.QUuid {
        return qtc.QUuid_new4(l, w1, w2, b1, b2, b3, b4, b5, b6, b7, b8);
    }


    /// New5 constructs a new QUuid object.
    pub fn New5(id128: QtC.QUuid__Id128Bytes) QtC.QUuid {
        return qtc.QUuid_new5(@ptrCast(id128));
    }


    /// New6 constructs a new QUuid object.
    pub fn New6(guid: ?*anyopaque) QtC.QUuid {
        return qtc.QUuid_new6(@ptrCast(guid));
    }


    /// New7 constructs a new QUuid object.
    pub fn New7(param1: ?*anyopaque) QtC.QUuid {
        return qtc.QUuid_new7(@ptrCast(param1));
    }


    /// New8 constructs a new QUuid object.
    pub fn New8(id128: QtC.QUuid__Id128Bytes, order: i32) QtC.QUuid {
        return qtc.QUuid_new8(@ptrCast(id128), @intCast(order));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QUuid_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QUuid_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn ToString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUuid_ToString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("quuid.ToString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ToByteArray(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QUuid_ToByteArray(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("quuid.ToByteArray: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn ToBytes(self: ?*anyopaque, ) QtC.QUuid__Id128Bytes {
        return qtc.QUuid_ToBytes(@ptrCast(self));
    }

    pub fn ToRfc4122(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QUuid_ToRfc4122(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("quuid.ToRfc4122: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn FromBytes(bytes: ?*anyopaque) QtC.QUuid {
        return qtc.QUuid_FromBytes(@ptrCast(bytes));
    }

    pub fn FromRfc4122(param1: []const u8) QtC.QUuid {
        return qtc.QUuid_FromRfc4122(@ptrCast(param1));
    }

    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QUuid_IsNull(@ptrCast(self));
    }

    pub fn OperatorAssign(self: ?*anyopaque, guid: ?*anyopaque) void {
        qtc.QUuid_OperatorAssign(@ptrCast(self), @ptrCast(guid));
    }

    pub fn Operator_GUID(self: ?*anyopaque, ) QtC._GUID {
        return qtc.QUuid_Operator_GUID(@ptrCast(self));
    }

    pub fn CreateUuid() QtC.QUuid {
        return qtc.QUuid_CreateUuid();
    }

    pub fn CreateUuidV5(ns: QtC.QUuid, baseData: []const u8) QtC.QUuid {
        return qtc.QUuid_CreateUuidV5(@ptrCast(ns), @ptrCast(baseData));
    }

    pub fn CreateUuidV3(ns: QtC.QUuid, baseData: []const u8) QtC.QUuid {
        return qtc.QUuid_CreateUuidV3(@ptrCast(ns), @ptrCast(baseData));
    }

    pub fn Variant(self: ?*anyopaque, ) i32 {
        return qtc.QUuid_Variant(@ptrCast(self));
    }

    pub fn Version(self: ?*anyopaque, ) i32 {
        return qtc.QUuid_Version(@ptrCast(self));
    }

    pub fn Data1(self: ?*anyopaque, ) u32 {
        return qtc.QUuid_Data1(@ptrCast(self));
    }

    pub fn SetData1(self: ?*anyopaque, data1: u32) void {
        qtc.QUuid_SetData1(@ptrCast(self), data1);
    }

    pub fn Data2(self: ?*anyopaque, ) u16 {
        return qtc.QUuid_Data2(@ptrCast(self));
    }

    pub fn SetData2(self: ?*anyopaque, data2: u16) void {
        qtc.QUuid_SetData2(@ptrCast(self), data2);
    }

    pub fn Data3(self: ?*anyopaque, ) u16 {
        return qtc.QUuid_Data3(@ptrCast(self));
    }

    pub fn SetData3(self: ?*anyopaque, data3: u16) void {
        qtc.QUuid_SetData3(@ptrCast(self), data3);
    }

    pub fn OperatorAssign2(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QUuid_OperatorAssign2(@ptrCast(self), @ptrCast(param1));
    }

    pub fn ToString1(self: ?*anyopaque, mode: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUuid_ToString1(@ptrCast(self), @intCast(mode));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("quuid.ToString1: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ToByteArray1(self: ?*anyopaque, mode: i32, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QUuid_ToByteArray1(@ptrCast(self), @intCast(mode));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("quuid.ToByteArray1: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn ToBytes1(self: ?*anyopaque, order: i32) QtC.QUuid__Id128Bytes {
        return qtc.QUuid_ToBytes1(@ptrCast(self), @intCast(order));
    }

    pub fn FromBytes2(bytes: ?*anyopaque, order: i32) QtC.QUuid {
        return qtc.QUuid_FromBytes2(@ptrCast(bytes), @intCast(order));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QUuid_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/quuid-id128bytes.html
pub const quuid__id128bytes = struct {

    /// New constructs a new QUuid::Id128Bytes object.
    pub fn New() QtC.QUuid__Id128Bytes {
        return qtc.QUuid__Id128Bytes_new();
    }


    /// New2 constructs a new QUuid::Id128Bytes object.
    pub fn New2(param1: ?*anyopaque) QtC.QUuid__Id128Bytes {
        return qtc.QUuid__Id128Bytes_new2(@ptrCast(param1));
    }


    pub fn OperatorQByteArrayView(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QUuid__Id128Bytes_OperatorQByteArrayView(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("quuid::id128bytes.OperatorQByteArrayView: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QUuid__Id128Bytes_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/uuid.html#types
pub const enums = struct {
    pub const Variant = enum {
        pub const VarUnknown: i32 = -1;
        pub const NCS: i32 = 0;
        pub const DCE: i32 = 2;
        pub const Microsoft: i32 = 6;
        pub const Reserved: i32 = 7;
    };

    pub const Version = enum {
        pub const VerUnknown: i32 = -1;
        pub const Time: i32 = 1;
        pub const EmbeddedPOSIX: i32 = 2;
        pub const Md5: i32 = 3;
        pub const Name: i32 = 3;
        pub const Random: i32 = 4;
        pub const Sha1: i32 = 5;
    };

    pub const StringFormat = enum {
        pub const WithBraces: i32 = 0;
        pub const WithoutBraces: i32 = 1;
        pub const Id128: i32 = 3;
    };

};
